src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts

```diff
@@ -0,0 +1,217 @@
+        var marginBoxParent = $(element).parents('.marginBox');
+        var parentBottom;
+        if(marginBoxParent && marginBoxParent.length > 0)
+            parentBottom = $(marginBoxParent[0]).offset().top + $(marginBoxParent[0]).outerHeight(true);
+        else
+            parentBottom = 999999;
+        var elemTop = $(element).offset().top;
+        var elemBottom = elemTop + $(element).outerHeight(false);
+        // console.log("Offset top: " + elemTop + " Outer Height: " + $(element).outerHeight(false));
+        // If css has "overflow: visible;", scrollHeight is always 2 greater than clientHeight.
+        // This is because of the thin grey border on a focused input box.
+        // In fact, the focused grey border causes the same problem in detecting the bottom of a marginBox
+        // so we'll apply the same 'fudge' factor to both comparisons.
+        var focusedBorderFudgeFactor = 2;
+
+        return  elemBottom > parentBottom + focusedBorderFudgeFactor;
+    }
+
+    // Checks for overflow on a bloom-page and adds/removes the proper class
+    // N.B. This function is specifically designed to be called from within AddOverflowHandler()
+    // but is also called from within StyleEditor (and therefore public)
+    public static MarkOverflowInternal(box) {
+        // There are two types of overflow that we need to check.
+        // 1-When we're called by a handler on an element, we need to check that that element
+        // doesn't overflow internally (i.e. has too much stuff to fit in itself).
+        // 2-We also need to check that this element and any OTHER elements on the page
+        // haven't been pushed outside the margins
+
+        // Type 1 Overflow
+        var $box = $(box);
+        $box.removeClass('overflow');
+        if (OverflowChecker.IsOverflowingSelf(box)) {
+            $box.addClass('overflow');
+        } else {
+            //now, thing is, while the text may fit in our box, our box may not fit our parent. Or grandparent, etc.
+            //It could be that we could just do this on up the hierarchy? For now, here's the case we know is important,
+            // in the Origami pages, where the space allocated could be too small.
+            if (OverflowChecker.HasSplitParentThatOverflows($box)) {
+                $box.addClass('overflow');
+            }
+        }
+
+        var container = $box.closest('.marginBox');
+        //NB: for some historical reason in March 2014 the calendar still uses textareas
+        var queryElementsThatCanOverflow = ".bloom-editable, textarea";
+        var editablePageElements = $(container).find(queryElementsThatCanOverflow);
+
+        // Type 2 Overflow - We'll check ALL of these for overflow past marginBox
+        editablePageElements.each(function() {
+            var $this = $(this);
+            if (OverflowChecker.IsOverflowingMargins($this[0])) {
+                $box.addClass('overflow'); // probably typing in the focused element caused this
```

Comment:
That's what the 'else' just below this is for (at least partially). If they go and fix the overflow elsewhere, then $box will fall into the category of the else when the edited box is processed for overflow and its overflow marker will be removed there.